File: essdee_attendance/essdee_attendance/report/work_operation_report/work_operation_report.py
import frappe
from frappe import _

def execute(filters=None):
	columns, data = [],[]
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_columns(filters):
	columns = [
		{'fieldname': 'employee_id', 'fieldtype': 'Data', 'label': 'Employee ID', 'options': 'Employee','width':180},
		{'fieldname': 'employee_name', 'fieldtype': 'Data', 'label': 'Employee Name','width':180},
		{'fieldname': 'operation', 'fieldtype': 'Data', 'label': 'Operation','width':180},
		{'fieldname': 'quantity', 'fieldtype': 'Int', 'label': 'Quantity','width':100},
		{'fieldname': 'total', 'fieldtype': 'Currency', 'label': 'Total','width':180}
	]
	if not filters.summarized_view:
		columns.insert(3, {'fieldname': 'rate', 'fieldtype': 'Currency', 'label': 'Rate','width':100})
		columns.insert(5, {'fieldname': 'uom', 'fieldtype': 'Data', 'label': 'UOM','width':180})
	return columns

def get_data(filters):
	data = []
	Essdee_Work_Entry = frappe.qb.DocType('Essdee Work Entry')
	Essdee_Work_Entry_Detail = frappe.qb.DocType('Essdee Work Entry Detail')
	Employee = frappe.qb.DocType('Employee')
	query = frappe.qb.from_(Essdee_Work_Entry).join(Essdee_Work_Entry_Detail) \
		.on(Essdee_Work_Entry.name == Essdee_Work_Entry_Detail.parent)
	
	query = query.join(Employee).on(Essdee_Work_Entry.employee == Employee.name)

	query = query.select(
			Employee.name,
			Employee.employee_name,
			Essdee_Work_Entry_Detail.operation,
			Essdee_Work_Entry_Detail.rate,
			Essdee_Work_Entry_Detail.quantity,
			Essdee_Work_Entry_Detail.uom,
			Essdee_Work_Entry_Detail.total,
		).where(Essdee_Work_Entry.date.between(filters.from_date, filters.to_date))
	
	if filters.employee:
		query = query.where(Essdee_Work_Entry.employee == filters.employee)
	if filters.employment_type:
		query= query.where(Employee.employment_type == filters.employment_type)
	if filters.department:
		query= query.where(Employee.department == filters.department)
	if filters.default_shift:
		query= query.where(Employee.default_shift == filters.default_shift)
	if filters.branch:
		query= query.where(Employee.branch == filters.branch)

	query = query.orderby(Employee.name)			 
	
	result = query.run(as_dict = True)

	for row in result:
		employee_details = {
			'employee_id': row.name,
			'employee_name': row.employee_name,
			'operation': row.operation,
			'rate': row.rate,
			'quantity': row.quantity,
			'uom':row.uom,
			'total': row.total
		}
		if not filters.summarized_view:
			existing_entry = next((item for item in data if item['employee_id'] == employee_details['employee_id'] and item['operation'] == employee_details['operation']), None)
			if existing_entry:
				existing_entry['quantity'] += employee_details['quantity']
				existing_entry['total'] = existing_entry['quantity']*existing_entry['rate']
			else:
				data.append(employee_details)
		else:
			employee_details.pop('rate')
			existing_entry = next((item for item in data if item['employee_id'] == employee_details['employee_id']), None)
			if existing_entry:
				existing_entry['quantity'] += employee_details['quantity']
				existing_entry['total'] += employee_details['total']
				if employee_details['operation'] not in existing_entry['operation']:
					existing_entry['operation'] += "," + employee_details['operation']
			else:
				data.append(employee_details)
	return data

File: essdee_attendance/essdee_attendance/report/work_operation_report/work_operation_report.js
// Copyright (c) 2024, Aerele and contributors
// For license information, please see license.txt

frappe.query_reports["Work Operation Report"] = {
	"filters": [
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.add_days(frappe.datetime.week_start(), -2)
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.add_days(frappe.datetime.week_end(), -2)
		},
		{
			"fieldname":"employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			get_query: () => {
				var company = frappe.query_report.get_filter_value('company');
				// var employment = frappe.query_report.get_filter_value('employment_type')
				return {
					filters: {
						'company': company,
					}
				};
			}
		},
		{
			"fieldname":"employment_type",
			"label": __("Employment Type"),
			"fieldtype": "Link",
			"options": "Employment Type",
		},
		{
			"fieldname":"default_shift",
			"label": __("Shift"),
			"fieldtype": "Link",
			"options": "Shift Type",
		},
		{
			"fieldname":"branch",
			"label": __("Branch"),
			"fieldtype": "Link",
			"options": "Branch",
		},
		{
			"fieldname":"department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department",
		},
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"summarized_view",
			"label": __("Summarized View"),
			"fieldtype": "Check",
			"default": 0,
			on_change: () => {
				frappe.query_report.refresh();
			}
		},
	]
};
